Per-pool timezone support

motiejus opened the pull request:
This pull request consists of 2 commits:
1. Logically separate emysql:add_pool API to two stages: required and optional
2. Add per-pool time_zone support

First commit also makes much easier to add other per-pool parameters.

This pull request is backwards compatible.

Supersedes #42.

Commit:
Separate Optional stuff from emysql:add_pool

Logically separate emysql:add_pool API to two stages:
1. Connection establishment (connect to the database server)
2. Miscellaneous options (select database, select encoding)

Now emysql:add_pool/6 takes 6 arguments that are absolutely necessary to
create a pool:
    pool name, number of connections, username, password, host, port.

emysql:add_pool/7 might take a proplist as last argument of optional fields:
    database, encoding

This makes adding optional pool settings much easier without adding more
and more arguments to emysql:add_pool.

Note: Since current codebase heavily depends on knowledge of encoding,
if encoding is not specified, 'utf8' is chosen by default (configurable
in emysql.hrl).

Commit:
Add per-pool time_zone support

bvdeenen commented:
It seems to me the change is completely backward compatible and should not impact anything once the connection has been established? Could you please comment on that. 
If so, I'm for merging (which I can do, but I still consider it somewhat more eonblasts code).
Bart.

motiejus commented:
Yes, patch is fully backwards compatible. There is no change if old `emysql:add_pool/8` is used.

This pull request changes only start-up behavior of _newly created pools_, and only when they are called with the _new_ `emysql:add_pool/7`.

motiejus commented:
... unless somebody (ehem, SSP) depends on `#pool{}` record (which is an internal record and is supposed to be opaque) then it is not backwards compatible. Then of course it has to be adjusted. But adjustments are minor.

yfyf commented:
I think it's good for merging. The `#pool{}` stuff you really shouldn't worry about. At first I thought that `database` was a mandatory option originally, but it seems that you were allowed to pass `undefined`, so you are keeping the original semantics by switching to proplists. And I think it's good to make encoding optional and default to 'utf8'.

hdiedrich commented:
Hi Motiejus,

excellent stuff, thanks for contributing!

There are actually the test cases you added beyond the options and timezone settings which makes it almost three different contributions.

I am positive 70%+ of your proposals should end up in the driver very soon now, I am not convinced about making the database an option. I can see two reasons pro: being able to have a different arity function call for the new API layout and because of the internal set_\* function that can be fed the database just like any other option, internally. There is even a third reason: as I just learned e.g. Sphynx prefers to have no database set at all which clearly makes setting the database 'an option' in that case (if one that is not available).

However that, in my view, does not justify to take the database out of the "core" parameters, which it is in most cases, with a normal MySQL server. Leaving the database name out, in my experience, would be the rare thing to do.

Instead, a guard checking for a list in the last parameter to distinguish it from the encoding parameter might be a better choice. The elegance of the set_option function would be violated by the special case of the database but that, in my view, would be a price worth paying in the light of what I outlined above. Another small merge of today allows for empty database parameters for Sphynx.

If you could bring yourself to rework your contribution along these lines, and make a new pull request, fantastic. If not, I will do it at some point and take care to preserve your lines, and your blame credit as much as possible.

Again, thanks for posting, let me know what you think -- Bart, your opinion to my thoughts is very much appreciated, too.

Thanks,
Henning

motiejus commented:
I am happy to rework the patch and agree with the suggestion. However, what kind of exact API would you see fit?
One of the options:
1. `add_pool(PoolId, Size, User, Password, Host, Port, Opts)` <- currently proposed API
2. `add_pool(PoolId, Size, User, Password, Host, Port, Database, Opts)` <- you are proposing
3. `add_pool(PoolId, Size, User, Password, Host, Port, Database)` <- for convenience in your proposed API
4. `add_pool(PoolId, Size, User, Password, Host, Port, Database, Encoding)` <- backwards compatibility

Technically, I would `Opts` from `Database` and `Encoding` would be separated by type: `Opts` would be a proplist (a pattern match when first element of the list is a tuple), whereas `Database` and `Encoding` would be strings (`is_list/1 orelse is_binary/1`).

Is that suitable?

Eonblast commented:
Hi Motiejus!

This is great. I meant all of 2,3 and 4. Holding 2 and 4 apart by a guard, Encoding has to be an atom, Opts a list.

I re-checked with tests and docs, Encoding is not a list, it's always supposed to be an atom as far as I can see.

The only incompatibility would then be "forward" with your #1 and the convenience #3, the Database name and Opts are both lists. If we don't enforce Opts to be a proplist I imagine that this would cause changes that break your current use of your fork of Emysql. Is that tolerable for you, are those few enough places that you can change the calls so we could drop #1?

If so, then Opts would not need to be a proplist, which would allow for single-atom options in the future, which I find very valuable. Concretely there are speed and fringe-case handling cases that I am currently looking at that can use single-atom options.

Thanks for your thoughts,
Henning

motiejus commented:
Hi,

OK, I will discard option 1, it is not a problem from our side. In practice everybody needs to select `Database`, but in those rare cases we can pass `undefined` anyway.

I will submit a commit updating this later today. Thank you for your comments.

hdiedrich commented:
Great, thanks!

If that concerns the patch, please allow for "" and <<>> besides 'undefined' for "dont't make a database selection".

Best,
Henning

Commit:
Mandatory `Database` option in emysql:add_pool API

It can be empty ('undefined', "" or <<"">>) anyway, but the use of MySQL
without a selected database is too limited to be practically considered
in the API.

motiejus commented:
API changed like we agreed.

hdiedrich commented:
Great stuff, I am half way through reviewing, will get back to it later while I am merging smaller change sets, to keep the diff noise low.

Thanks again for contributing!

Henning

Commit:
NEW sql_mode

jlouis commented:
I skimmed these and I think they look good from a quick look. I did not read through them with a teaspoon in my hand for nitpicking however.

jlouis commented:
@Motiejus Do you still have an interest in this one? I am pretty confident I like the code, but I did some major surgery on the internals so it may be that we need to do some serious rebase-work on this branch to make it mergeable again. Before I embark on that however, I want to know if there is still a need for this.

You can also give it a try yourself, if you feel so inclined.

motiejus commented:
Jesper,

it is relevant. We are using our fork with this patch internally in Spil Games.

I would be very glad if you looked at this, as I am now on holiday. However, if you would rather not, I will gladly take care of this once I am back, roughly after 3-4 weeks.

Note that after a while of using it I have another idea: instead of custom options proplist, pass arbitrary list of binaries to emysql:add_pool/X, which would be executed on connection start-up on every socket. At the moment we set:

``` erlang
Options = [
    {encoding, utf8},
    {time_zone, "+00:00"},
    {sql_mode, "STRICT_ALL_TABLES"}
],
```

Arbitrary list of binaries could save us code in emysql and would give flexibilty for the user change internal state of each connection on start-up.

Whichever you prefer.

jlouis commented:
Thank you, and have a nice vacation!
